File: typescript/ch5/Just.ts
import { Maybe } from './Maybe';
import { MaybeFactory } from './MaybeFactory';
import { WrapperFunc } from './WrapperFunc';

class Just extends Maybe {
  protected _value: any;

  constructor(value: any) {
    super();
    this._value = value;
  }

  get value() {
    return this._value;
  }

  map(f: WrapperFunc) {
    return MaybeFactory.fromNullable(f(this._value));
  }

  getOrElse(other?: any) {
    return this._value;
  }

  filter(f: WrapperFunc) {
    MaybeFactory.fromNullable(f(this._value) ? this._value : null);
  }

  chain(f: WrapperFunc) {
    return f(this._value);
  }

  toString() {
    return `Maybe.Just(${this._value})`;
  }

  get isJust() {
    return true;
  }
}

export {
  Just
};

File: python/python/ch5/Just.py
from .Maybe import Maybe
from .MaybeFactory import MaybeFactory

class Just(Maybe):
    def __init__(self, value):
        self._value = value
        
    @property
    def value(self):
        return self._value
      
    @property
    def is_just():
        return True
      
    def map(self, f):
        return MaybeFactory.from_nullable(f(self._value))
      
    def get_or_else(self, other=None):
        return self._value
      
    def filter(self, f):
        MaybeFactory.from_nullable(self._value if f(self._value) else None)
        
    def chain(self, f):
        return f(self._value)
      
    def __str__(self):
        return f'Maybe.Just({self._value})'